llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -1376,10 +1377,32 @@ static void computeKnownBitsFromOperator(const Operator *I,
       for (unsigned i = 0; i != SubScale; ++i) {
         computeKnownBits(I->getOperand(0), SubDemandedElts.shl(i), KnownSrc, Q,
                          Depth + 1);
-        unsigned ShiftElt = Q.DL.isLittleEndian() ? i : SubScale - 1 - i;
+        unsigned ShiftElt = IsLE ? i : SubScale - 1 - i;
         Known.insertBits(KnownSrc, ShiftElt * SubBitWidth);
       }
     }
+    // Look through a cast from wider vector elements to narrow type.
+    // Examples: v2i64 -> v4i32
+    if (SubBitWidth % BitWidth == 0) {
```

Comment:
I guess the missing case `v2i32 -> i64` can be handled here (as a follow-up): https://github.com/llvm/llvm-project/blob/07fa6d1d90c714fa269529c3e5004a063d814c4a/llvm/lib/Analysis/ValueTracking.cpp#L1295-L1301